### PR TITLE
Feature/v2land

### LINKED ISSUE #653: Upgrade Python converter for land to IODA-v2: Use the IODA-v2 API
Update the converter identified in the title (hereafter referred to as the **target converter**) to use the IODA-v2 API.

- Update import statement: from 
- Variable dimensions dictionary
- Dimensions dictionary
- Global attributes dictionary
- Remove record numbers
- Update keyDict
- Units, Fill values, other metadata
- Change order of when writer is called
- Remove ExtractoObsData if possible
- Update CMakeLists.txt with environment variable
- Run test to make sure it runs (but will fail because of unupdated test reference)
- Update test reference file
- run unit tests making use of the target converter and verify that they pass. Make sure you verify the results by examining the file using ```ncdump```, ```h5dump```, ```HDFview```, or a similar utility.

Converters are included below:

- afwa_snod2ioda.py
- ascat_ssm2ioda.py
- ghcn_snod2ioda.py
- ims_scf2ioda.py
- smap_ssm2ioda.py
- smos_ssm2ioda.py

## ZenHub organization

Please make sure that you have properly set:
- Labels: OBS3
- Milestone: OBS - October 2021
- Estimate: An estimate of how long this issue will take to resolve. 1 = half a day. 2 = full day.
- Epics: Modify converters to IODAv2 format
- Assignees: Anyone **working** on this issue. Anyone who is not working directly on the issue, but
  who should receive notifications, should be listed below.
- Reviewers: 

## Notifications

Fill in a list of GitHub handles here. People here are not *assigned*, but instead receive notifications
about this issue's progress.

Notifications should always go to: @rhoneyager.

Other people:

### PR DESCRIPTION
**Pull request by YoulongXia-NOAA**
## Description

* Upgrade ioda v2 from v1 with the ioda engine

Use newly developed ioda engine to upgrade 6 python scripts in land directory from v1 to v2 as the code-sprint instructed.

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #653

## Acceptance Criteria (Definition of Done)

Ctest will be passed and merged into develop branch

## Dependencies

No

## Impact

No

### CODE REVIEW

**Review by srherbener — APPROVED** [review]
Looks good. Thanks for your patience as we worked out the specifications for the units.

**Comment by srherbener** [comment]
@rhoneyager have all of your comments been addressed? Thanks!

**Comment by YoulongXia-NOAA** [comment]
I did not update nlocs yet. I need to git clone and merge develop to my local branch.  Please wait for a  bit time.

**Comment by YoulongXia-NOAA** [comment]
@srherbener, I will wait for further instruction about how to set up nlocs to zero. Thank you.

**Comment by srherbener** [comment]
@YoulongXia-NOAA, the remaining test failures are due to the reference files in `iodaconv/test/testoutput` containing 1..n values in the nlocs variable instead of all zeros (which is now expected due to yesterday's PR merges). You should be able to copy the files from `your_build_area/iodaconv/test/testrun` to `iodaconv/test/testoutput` to get these back in sync.

**Comment by YoulongXia-NOAA** [comment]
> @YoulongXia-NOAA, the remaining test failures are due to the reference files in `iodaconv/test/testoutput` containing 1..n values in the nlocs variable instead of all zeros (which is now expected due to yesterday's PR merges). You should be able to copy the files from `your_build_area/iodaconv/test/testrun` to `iodaconv/test/testoutput` to get these back in sync.

Do I need to git pull from feature/v2land to rerun my ctest or I need to merge develop to my feature/v2land in local repository and rerun ctest to produce new output in testrun directory, which way do you suggest?

**Comment by srherbener** [comment]
You can do the update in your local repository. Re-run the tests and copy the new output to the test reference.
```
# using your feature/v2land branch
cd your_build_area
make update
make -j4
ctest -R iodaconv

cp <your_build_area>/iodaconv/test/testrun/<your_file> <your_source_area>/ioda-bundle/iodaconv/test/testoutput/<your_file>
# repeat cp command for all of your files
```

**Comment by YoulongXia-NOAA** [comment]
@srherbener, thank you for your instruction. I have put it to my note. When I do: git checkout develop, git pull, git checkout feature/v2land, git merge, ecbuild, make -j4, and ctest -R ioda", the procedure is more complex. Such a method seems simple and convenient. could you please give me more explanations about which case I can use make update or which case I need to use the complex way. This will help me for the converter development.  I have updated my outputs and it passed.

**Comment by srherbener** [comment]
> @srherbener, thank you for your instruction. I have put it to my note. When I do: git checkout develop, git pull, git checkout feature/v2land, git merge, ecbuild, make -j4, and ctest -R ioda", the procedure is more complex. Such a method seems simple and convenient. could you please give me more explanations about which case I can use make update or which case I need to use the complex way. This will help me for the converter development. I have updated my outputs and it passed.

@YoulongXia-NOAA, the method I outlined above is for the case when you have already merged develop into your feature branch and you simply need to update the test reference files.

The method you outlined is for the case when you need to merge develop into your feature branch (which of course takes more steps).

**Comment by YoulongXia-NOAA** [comment]
Nice explanation, thank you @srherbener.

File: src/land/ghcn_snod2ioda.py [hunk 1]
```diff
@@ -159,16 +168,19 @@ def assignValue(colrowValue, df400):
...
             times[i] = base_datetime
-        self.loc_mdata['datetime'] = self.writer.FillNcVector(times, "datetime")
-        self.loc_mdata['latitude'] = lats
-        self.loc_mdata['longitude'] = lons
-        self.loc_mdata['altitude'] = alts
+
+        self.outdata[('datetime', 'MetaData')] = times
+        self.outdata[('latitude', 'MetaData')] = lats
+        self.outdata[('longitude', 'MetaData')] = lons
+        self.outdata[('altitude', 'MetaData')] = alts
```
[the remark this answers is not in the document]
Comment (srherbener) [reply src/land/ghcn_snod2ioda.py:175]: @rhoneyager, I just built and tested on my mac. All tests pass, and the output `ghcn_snod_20200228.nc` file does indeed contain these variables with the proper units and fill values. Does this address your comment? Thanks!


Screen Shot 2021-11-03 at 9 05 20 AM

**Review by rhoneyager — APPROVED** [review]
It looks good. Thanks!

**Pull request merged**